lib/debug/server_dap.rb

```diff
@@ -687,7 +687,23 @@ def register_vars vars, tid
   class ThreadClient
     def value_inspect obj
       # TODO: max length should be configuarable?
-      DEBUGGER__.safe_inspect obj, short: true, max_length: 4 * 1024
+      value = DEBUGGER__.safe_inspect obj, short: true, max_length: 4 * 1024
+
+      # Given that this is going to be transmitted in a JSON string, it needs to be unicode-encoded
+      # (and "UTF-8" is the default).
+      if value.encoding != Encoding::UTF_8
+        # If the string we got is frozen, we need to make a copy first
+        value = value.dup if value.frozen?
+        value.force_encoding(Encoding::UTF_8)
+      end
+
+      if value.valid_encoding?
+        value
+      else
+        # If a variable contains non-unicode data, at least we can send it partially and signal that
+        # the encoding was unexpected.
+        "[Invalid encoding] #{value.encode("UTF-8", invalid: :replace, undef: :replace)}"
+      end
```

Comment:
```suggestion
      str = DEBUGGER__.safe_inspect obj, short: true, max_length: 4 * 1024

      if str.encoding == Encoding::UTF_8
        str.scrub
      else
        str.encode(Encoding::UTF_8, invalid: :replace, undef: :replace)
      end
```

is enough. could you check it?

Reply:
This might solve the UTF-8 problem, but (in order of importance):

1. Doesn't take into account the frozen string problem
2. Doesn't have any comments explaining to future developers *why* we are doing the manipulation
3. I don't believe this is the right thing to do from a UX standpoint: This gives no indication to the user that what they are seeing is *not* the data (which the `[Invalid encoding]` prefix tries to communicate)

Reply:
1. could you give me an example? it doesn't modify the generated string.
2. IMO the code is obvious, and `git blame` is enough.
3. `[Invalid encoding]` is too much because it is ambiguous what is "invalid". JSON doesn't support non-UTF-8 string but the program is valid (the program isn't invalid code). And most of case `inspect` returns valid UTF-8 encoding, I don't think we need care more.

Reply:
1. You are right, just double-checked with the test I added for frozen strings, and it works

The other 2 are more of a personal/UX preference, so I defer to you, as the maintainer.

IMO, it's not obvious for someone reading the code without context/without knowing that JSON doesn't support non-unicode strings (the specs don't talk about UTF-8).

I'm accepting the proposed changes, and we can think about better documenting in the future if needed